### PR TITLE
arch/arm64: Fix clock drift from tick timer

### PR DESCRIPTION
**Pull request by Fix-Point**
## Summary
In the previous implementation, when the current timer count just exceeds the next tick time, it will still cause clock drift. This commit fixed clock drift by using the abosolute ticks (clock_systime_tick() + ticks).

## Impact
Fixed timing errors on ARMv8A platform.

## Testing
Tested on QEMU/armv8a with the test case in `https://github.com/apache/nuttx/pull/13674`.

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
@jlaitine could you review and try this patch?

**Review by xiaoxiang781216 — APPROVED** [review]

File: arch/arm64/src/common/arm64_arch_timer.c [hunk 1]
```diff
@@ -286,9 +286,8 @@ static int arm64_tick_start(struct oneshot_lowerhalf_s *lower,
 
   priv->running = this_cpu();
 
-  next_cycle =
-    arm64_arch_timer_count() / priv->cycle_per_tick * priv->cycle_per_tick +
-    ticks * priv->cycle_per_tick;
+  next_cycle = (clock_systime_ticks() + ticks) *
```
Comment (jlaitine) [inline arch/arm64/src/common/arm64_arch_timer.c:289]: This puts bsck the timer drift; there is an unknown delay between tick interrupt taken and this one called, which caises the drift.

The current code fixes the drift issue by flooring the current time to an even multiple of tick time; thus also the next_cycle is at even multiple of tick time, and each cycle is exactly of correct length. Also querying the current tick always then returns the same number during the tick period.

**Review by jlaitine — COMMENTED** [review]
I don't understand this patch, it is just putting back the issues which were just fixed. Is there some problem which this is trying to solve?

**Comment by jlaitine** [comment]
Maybe post details of the issue and some test to re-produce it?

**Review by jlaitine — CHANGES_REQUESTED** [review]
I think this just breaks things more

**Comment by Fix-Point** [comment]
I am using this test case to reproduce the bug.

```c
#include <nuttx/nuttx.h>
#include <nuttx/config.h>
#include <stdio.h>

#define read_sysreg(reg)                         \
  ({                                             \
    uint64_t __val;                              \
    __asm__ volatile ("mrs %0, " STRINGIFY(reg)  \
                    : "=r" (__val) :: "memory"); \
    __val;                                       \
  })

static inline uint64_t arm64_arch_timer_count(void)
{
  return read_sysreg(cntvct_el0);
}

static inline uint64_t arm64_arch_timer_get_cntfrq(void)
{
  return read_sysreg(cntfrq_el0);
}

static inline uint64_t time_us(void)
{
  return arm64_arch_timer_count() * 1000000 / arm64_arch_timer_get_cntfrq();
}

inline uint64_t minutes_in_ticks(uint64_t mins)
{
  return mins * 60 * 1000000 / CONFIG_USEC_PER_TICK;
}

int main(int argc, FAR char *argv[])
{
  sem_t sem;
  uint64_t t1;
  uint64_t t2;
  sem_init(&sem, 0, 0);

  for (int64_t i = 0; i < minutes_in_ticks(10); i++)
    {
      t1 = time_us();
      nxsem_tickwait(&sem, 1);
      t2 = time_us();

      if (i % 100 == 0)
        {
          printf("%lu %lu %lu\n",t1,t2,t2-t1);
        }

      if (t2 - t1 < CONFIG_USEC_PER_TICK)
        {
          printf("ERROR: slept %lu us instead of minimum %u\n", t2-t1, CONFIG_USEC_PER_TICK);
        }

      assert(t2-t1 > CONFIG_USEC_PER_TICK); 
    }

  return 0;
}
```

1. The problem that tickwait_uninterruptible may wake up one tick too early is caused by the clock drift. I tried the following scenarios:
- Without this patch and delay + 1, bug reproduced.
- With this patch, bug fixed.
- Without https://github.com/apache/nuttx/pull/13674, bug reproduced.
- Without https://github.com/apache/nuttx/pull/13674 but with this patch, bug fixed.

2. The previous method of aligning down to ticks has a problem. If the current count obtained when setting the timer is aligned down and is 1 larger than the current system tick, it will still cause clock drift. The key is not to use the count obtained when setting the timer as a basis for delay, instead we should use current system tick.

**Comment by jlaitine** [comment]
What is the +1 bug you are referring to? Are you getting an error print and assertion with the test case above?
Re-starting the timer after systick has already passed means that servicing the tick interrupt took longer than the tick. This would be fatal in many ways, but I agree that the case is not handled in tick timer. However, the solution is not to add drift by reading the current time in the isr and adding the tick time to that. This would just cause clock drift; as I explained earlier.

I'll have a closer look later today!

**Comment by jlaitine** [comment]
Now I am able to re-produce the issue you are reporting in qemu, but not in real HW.

The issue in qemu is, that sometimes servicing the timer interrupt is delayed until close to the end of the next tick, or past it. This issue must  be related to real-time behaviour of qemu.

The same would happen also in a real HW, in case of interrupt storm or in case of very long critical section, delaying the timer interrupt.

There is no way around that issue! In case servicing the timer interrupt is late, it just MUST proceed several ticks at once to keep up. The same would happen if you e.g. break the target with a debugger, and the timer is freely running in the background.

Sorry to say, but I believe this patch is not right. As I explained before, every time you set the timer comparator, you must set the next value to be exactly the previous_comparator_time + tick_time. So you need to know the previous time, which was set somehow.

If you do as suggested in this patch, setting it to current_time() + tick time, you add a delay of
`deltaT = current_time() - previous_time; `
to each tick, causing timer drift.

However, when inspecting the full flow, I did find one race condition, which could cause an error. This is in theory, so unlikely that testing it is probably impossible without specifically planned instrumentation code...

This part in arm64_arch_timer.c

```
  next_cycle =
    arm64_arch_timer_count() / priv->cycle_per_tick * priv->cycle_per_tick +
    ticks * priv->cycle_per_tick;
 
  arm64_arch_timer_set_compare(next_cycle);

```

Has a potential race in the case that tick timer interrupt is missed (long due), and the new tick occurs between the lines of calculating next cycle and setting the compare, this could set the compare to the past.

Even though the possibility of this would be non-existent in practice, we can fix this easily if we want, by just checking the possibility that we didn't set the time in the past:

```
do
  {
    next_cycle =
      arm64_arch_timer_count() / priv->cycle_per_tick * priv->cycle_per_tick +
      ticks * priv->cycle_per_tick;
 
    arm64_arch_timer_set_compare(next_cycle);
} while(next_cycle - arm64_arch_timer_count() < priv->cycle_per_tick);  /* arm64_arch_timer_count() < next_cycle */

```
I will provide a patch for this. Even if it is highly theoretical case, it is probably worth fixing. This has nothing to do with the issue you are seeing in qemu, though.

**Comment by jlaitine** [comment]
A summary: I don't think that the tick timer can be reliably tested in qemu at all. The timer counting vs. interrupt latency is nowhere close to real-time or even predictable.

**Comment by Fix-Point** [comment]
Thank you for the comments. I am trying to make another patch to fix this problem.

**Pull request closed without merging**